// tsdav/dist/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ICALObjects = exports.DAVNamespaceShorthandMap = exports.DAVAttributeMap = exports.DAVNamespace = void 0;
var DAVNamespace;
(function (DAVNamespace) {
    DAVNamespace["CALENDAR_SERVER"] = "http://calendarserver.org/ns/";
    DAVNamespace["CALDAV_APPLE"] = "http://apple.com/ns/ical/";
    DAVNamespace["CALDAV"] = "urn:ietf:params:xml:ns:caldav";
    DAVNamespace["CARDDAV"] = "urn:ietf:params:xml:ns:carddav";
    DAVNamespace["DAV"] = "DAV:";
})(DAVNamespace = exports.DAVNamespace || (exports.DAVNamespace = {}));
exports.DAVAttributeMap = {
    [DAVNamespace.CALDAV]: 'xmlns:c',
    [DAVNamespace.CARDDAV]: 'xmlns:card',
    [DAVNamespace.CALENDAR_SERVER]: 'xmlns:cs',
    [DAVNamespace.CALDAV_APPLE]: 'xmlns:ca',
    [DAVNamespace.DAV]: 'xmlns:d',
};
exports.DAVNamespaceShorthandMap = {
    [DAVNamespace.CALDAV]: 'c',
    [DAVNamespace.CARDDAV]: 'card',
    [DAVNamespace.CALENDAR_SERVER]: 'cs',
    [DAVNamespace.CALDAV_APPLE]: 'ca',
    [DAVNamespace.DAV]: 'd',
};
var ICALObjects;
(function (ICALObjects) {
    ICALObjects["VEVENT"] = "VEVENT";
    ICALObjects["VTODO"] = "VTODO";
    ICALObjects["VJOURNAL"] = "VJOURNAL";
    ICALObjects["VFREEBUSY"] = "VFREEBUSY";
    ICALObjects["VTIMEZONE"] = "VTIMEZONE";
    ICALObjects["VALARM"] = "VALARM";
})(ICALObjects = exports.ICALObjects || (exports.ICALObjects = {}));
