// tsdav/dist-esm/util/requestHelpers.js
import { DAVAttributeMap, DAVNamespaceShorthandMap } from '../consts';
export const urlEquals = (urlA, urlB) => {
    if (!urlA && !urlB) {
        return true;
    }
    if (!urlA || !urlB) {
        return false;
    }
    const trimmedUrlA = urlA.trim();
    const trimmedUrlB = urlB.trim();
    if (Math.abs(trimmedUrlA.length - trimmedUrlB.length) > 1) {
        return false;
    }
    const strippedUrlA = trimmedUrlA.slice(-1) === '/' ? trimmedUrlA.slice(0, -1) : trimmedUrlA;
    const strippedUrlB = trimmedUrlB.slice(-1) === '/' ? trimmedUrlB.slice(0, -1) : trimmedUrlB;
    return urlA.includes(strippedUrlB) || urlB.includes(strippedUrlA);
};
export const urlContains = (urlA, urlB) => {
    if (!urlA && !urlB) {
        return true;
    }
    if (!urlA || !urlB) {
        return false;
    }
    const trimmedUrlA = urlA.trim();
    const trimmedUrlB = urlB.trim();
    const strippedUrlA = trimmedUrlA.slice(-1) === '/' ? trimmedUrlA.slice(0, -1) : trimmedUrlA;
    const strippedUrlB = trimmedUrlB.slice(-1) === '/' ? trimmedUrlB.slice(0, -1) : trimmedUrlB;
    return urlA.includes(strippedUrlB) || urlB.includes(strippedUrlA);
};
export const mergeObjectDupKeyArray = (objA, objB) => {
    return Object.entries(objA).reduce((merged, [currKey, currValue]) => {
        if (merged[currKey] && Array.isArray(merged[currKey])) {
            // is array
            return Object.assign(Object.assign({}, merged), { [currKey]: [...merged[currKey], currValue] });
        }
        if (merged[currKey] && !Array.isArray(merged[currKey])) {
            // not array
            return Object.assign(Object.assign({}, merged), { [currKey]: [merged[currKey], currValue] });
        }
        // not exist
        return Object.assign(Object.assign({}, merged), { [currKey]: currValue });
    }, objB);
};
export const getDAVAttribute = (nsArr) => nsArr.reduce((prev, curr) => (Object.assign(Object.assign({}, prev), { [DAVAttributeMap[curr]]: curr })), {});
export const formatProps = (props) => props === null || props === void 0 ? void 0 : props.reduce((prev, curr) => {
    var _a, _b;
    if (curr.namespace) {
        return Object.assign(Object.assign({}, prev), { [`${DAVNamespaceShorthandMap[curr.namespace]}:${curr.name}`]: (_a = curr.value) !== null && _a !== void 0 ? _a : {} });
    }
    return Object.assign(Object.assign({}, prev), { [`${curr.name}`]: (_b = curr.value) !== null && _b !== void 0 ? _b : {} });
}, {});
export const formatFilters = (filters) => filters === null || filters === void 0 ? void 0 : filters.map((f) => {
    var _a, _b;
    return ({
        [f.type]: Object.assign(Object.assign({ _attributes: f.attributes }, (_a = (f.children ? formatFilters(f.children) : [])) === null || _a === void 0 ? void 0 : _a.reduce((prev, curr) => mergeObjectDupKeyArray(prev, curr), {})), { _text: (_b = f.value) !== null && _b !== void 0 ? _b : '' }),
    });
});
export const cleanupFalsy = (obj) => Object.entries(obj).reduce((prev, [key, value]) => {
    if (value)
        return Object.assign(Object.assign({}, prev), { [key]: value });
    return prev;
}, {});
